src/translate.py

```diff
@@ -2232,42 +2260,34 @@ def add_imm(source: Expression, imm: Expression, stack_info: StackInfo) -> Expre
         # addiu $reg1, $reg2, 0 is a move
         # (this happens when replacing %lo(...) by 0)
         return source
-    elif source.type.is_pointer():
+    elif source.type.is_pointer_or_array():
         # Pointer addition (this may miss some pointers that get detected later;
         # unfortunately that's hard to do anything about with mips_to_c's single-pass
         # architecture.
-        if isinstance(imm, Literal):
+        # If the imm is exactly a multiple of 0x8000, then it is likely that we are
+        # constructing a large offset (see deref)
+        if isinstance(imm, Literal) and imm.value % 0x8000 != 0:
```

Comment:
hm, why doesn't the 0x7FFF case come up here? an example of it is in mm asm/overlays/ovl_Boss_07_0x809F4980.asm

Reply:
You're right, I decided to factor this out into `Literal.likely_partial_offset()`.
I think it wasn't an issue there because there wasn't a field at exactly 0x7FFF? (it's much more common for there to be a field at 0x8000?)

Reply:
I screwed up my testing of the last commit: it actually introduced a diff in OOT:
https://gist.github.com/zbanks/2e20069ef8f2d0e72b6b63c074a4c23d

All of the diffs from that change appear to be related to the `MessageContext` struct and `addiu ..., 0x7fff` instructions. The output seems completely improved.